src-self-hosted/Module.zig

```diff
@@ -2682,64 +2681,96 @@ fn analyzeInstElemPtr(self: *Module, scope: *Scope, inst: *zir.Inst.ElemPtr) Inn
     return self.fail(scope, inst.base.src, "TODO implement more analyze elemptr", .{});
 }
 
-fn analyzeInstSub(self: *Module, scope: *Scope, inst: *zir.Inst.Sub) InnerError!*Inst {
-    return self.fail(scope, inst.base.src, "TODO implement analysis of sub", .{});
+fn analyzeInstBinOp(self: *Module, scope: *Scope, inst: *zir.Inst) InnerError!*Inst {
+    return switch (inst.tag) {
+        .add => self.analyzeInstArithmetic(scope, inst.cast(zir.Inst.Add).?),
+        .sub => self.analyzeInstArithmetic(scope, inst.cast(zir.Inst.Sub).?),
+        else => self.fail(scope, inst.src, "TODO implement arithmetic for operand {}", .{@tagName(inst.tag)}),
+    };
 }
 
-fn analyzeInstAdd(self: *Module, scope: *Scope, inst: *zir.Inst.Add) InnerError!*Inst {
+/// Analyzes and performs the arithmetic operation specified by the given operand
+fn analyzeInstArithmetic(self: *Module, scope: *Scope, inst: anytype) InnerError!*Inst {
```

Comment:
I'm going to make a change in master branch that will allow this to be non generic - I hope you can bear with me. It will require rebasing your pull request, but the result will be this function not being generic and I think that's worth it.

The idea is to follow the pattern set so far by `ast.Node` and `ir.Inst` where multiple tags can map to the same type. This will make `add` and `sub` (and other arithmetic operations) have the same data structure, with only a differing enum tag.

I'll ping you when that's pushed to master branch, should be within a few hours from now.

Reply:
I'll gladly rework the PR once you've pushed your changes. Thank you for your patience with the PR so far.

Reply:
No problem! Thanks for the patch :)

I just pushed 8ee629aa4c7fb0127c71aec0b2c81353c4291ebb - can you try rebasing on top of that? It should simplify this patch, I think.